### DISCUSSION TITLE
How do I convert a 4bit PNG to 24bit PNG depth?

### DISCUSSION THREAD

**Discussion by Til-Weimann** [ImageMagick/ImageMagick#3453] [Help] [answered]
Hey,

I need to convert a lot of 4bit PNGs to 24bit while keeping the file format.

I have tried before:

1.  `convert img_4b.png -define png:bit-depth=24 img_24b.png` (and a couple of similar versions)
     => convert: ignoring invalid defined png:bit-depth =24 @ warning/png.c/WritePNGImage/12283.

2.  `convert img_4b.png -depth 24 img_24b.png`
     => no error, but returns a 4bit image / does nothing, works with 8bit though for some reason

Do you know a way I can do this?

Thanks in advance, 
Til

**Comment by fmw42** [comment]
Try prefacing the output with PNG24: such as PNG24:outimage.suffix

`convert img_4b.png PNG24:img_24b.png`

**Reply by Til-Weimann** [reply]
That works, thanks!

**Comment by fmw42** [comment, marked as answer]
For all PNG images in a directory, see mogrify at https://legacy.imagemagick.org/Usage/basics/#mogrify

`mogrify -format png PNG24:*.png`

This will overwrite all your PNGs. If you want to save to a pre-created empty folder, add `-path path_to_existing_folder`

I suggest you test on a folder with just a few images or make a backup folder before processing.

**Comment by tjex** [comment]
neither of these solutions work for me... :( 
Image stays at 8bit.

**Reply by fmw42** [reply]
Post and example 4bit image.  What is your IM version and platform/OS?

**Reply by tjex** [reply]
In my case it's 8bit. 
![tomatoes](https://user-images.githubusercontent.com/83029642/207099092-bbdb5ad0-1eeb-4359-b39f-f6640ecbec17.png)

IM: 7.1.0
OS: macos 12.6

I could get it to 16 bit with another solution found online `convert tomatoes.png -depth 16 PNG48:can-tomatoes.png`
But I can't deduce how to step it further up to 24bit...

**Reply by fmw42** [reply]
This image is 25bits. 8 bits for each of R,G,B and one bit for alpha.   What do you really want?  Please clarify.  Why did you mention 4-bits as input in  your title?

**Reply by tjex** [reply]
that's weird. Because when I checked with `identify can-tomatoes.png`, it showed me first 8 bits, and then 16 bits after I ran the above `convert` process. 

I need the file as a whole to be recognised as a 24bit png. I'm trying to upload to a service and it's throwing me the error that it is not a 24bit png.

Also, I'm not the op!

**Reply by fmw42** [reply]
Remove the alpha channel so it has only 24 bits

`convert can-tomatoes.png -alpha off -depth 8 new-can_tomatoes.png`

Perhaps they will accept it now.  Give it a try.

What is your IM version and platform/OS?

**Reply by tjex** [reply]
yep that did it! You're a legend, thanks for the prompt replies.
good lesson for me too, when writing error messages for users to say what it is about the file that is causing the error, and not just a multitude of options...
 
IM: 7.1.0
OS: macos 12.6

**Reply by fmw42** [reply]
FYI. -depth 8 means 8 bits for each of R,G,B so 24-bits.  You have 25 bits since the alpha had 1 bit and was pure opaque (no Transparency). We just had to remove the alpha.
